eqt/ui/UIFormWidget.py

```diff
@@ -320,25 +470,55 @@ def applyWidgetState(self, name, state, role=None):
                     widget.setChecked(value)
                 elif isinstance(widget, (QtWidgets.QTextEdit, QtWidgets.QPlainTextEdit)):
                     widget.setPlainText(value)
+                elif key == 'widget_number':
+                    if value != self.widget_number_dictionary[name]:
+                        self.widget_number_dictionary[name] = value
 
-    def applyWidgetStates(self, state):
+    def applyWidgetStates(self, states):
         '''
+        Removes the widgets in the form which are not present in the states. 
+        If the widgets in the states are not present in the form, 
+        they are retrieved from the removed-widgets dictionary and inserted at position 
+        given by the widget number recorded in the states. An error is raised when the 
+        widget in thet state is not in the form nor in the removed widgets.
         Applies the given states to the form's widgets.
 
         Parameters
         ----------
-        state: dict
-          Format: {'widget_name': {'value': str | bool | int, 'enabled': bool, 'visible': bool},
-                   ...},
-          e.g. {{'widget1': {'value': 1, 'enabled': True, 'visible': True},
-                 'widget2': {'value': 2, 'enabled': False, 'visible': False}}.
-        '''
-        for name, widget_state in state.items():
-            self.applyWidgetState(name, widget_state)
+        states: nested_dict
+          Format: {'name_field': {'value': str | bool | int, 'enabled': bool, 'visible': bool, 'widget_number' : int},
+                    'name_label': {'value': str | bool | int, 'enabled': bool, 'visible': bool}, ...},
+                  e.g. {{'widget1': {'value': 1, 'enabled': True, 'visible': True, 'widget_number' : int},
+                  'widget2': {'value': 2, 'enabled': False, 'visible': False}}.
+        '''
+        # remove extra widgets
+        if self.widgets.keys() > states.keys():
```

Comment:
What is this actually testing? You are comparing 2 lists.

Reply:
This  block of code + the next is comparing the current widgets in the form (which could contain less or extra widgets) to the states you are trying to retrieve. If the current widgets miss some widgets from the states they will be retrieved from the removed widget dictionary (in the next block of code). However if extra widgets were inserted, we must remove them and this is done here.

Reply:
I moved the remove extra widget block of code below the code to reinsert the eliminated widgets. In this way the > operator is used correctly.

Reply:
Note: The if statement with > returns "true" only if all of the elements in the list on the right are contained in the list of the left and the left list has extra elements too.